src/lib.rs

```diff
@@ -104,86 +105,127 @@ mod parser;
 const VERSION: &str = "0.0.1";
 const LANG: &str = "rust";
 
-impl Options {
-    /// Add a name option for the unconnected connection.
+#[doc(hidden)]
+pub mod options_typestate {
+    /// `Options` typestate indicating
+    /// that there has not yet been
+    /// any auth-related configuration
+    /// provided yet.
+    #[derive(Debug, Copy, Clone, Default)]
+    pub struct Unauthenticated;
+
+    /// `Options` typestate indicating
+    /// that auth-related configuration
+    /// has been provided, and may not
+    /// be provided again.
+    #[derive(Debug, Copy, Clone)]
+    pub struct Authenticated;
+
+    /// `Options` typestate indicating that
+    /// this `Options` has been used to create
+    /// a `Connection` and may not be changed.
+    #[derive(Debug, Copy, Clone)]
+    pub struct Finalized;
+}
+
+/// A configuration object for a NATS connection.
+#[derive(Clone, Debug, Default)]
+pub struct Options<Typestate = options_typestate::Unauthenticated> {
+    auth: AuthStyle,
+    name: Option<String>,
+    no_echo: bool,
+    typestate: PhantomData<Typestate>,
+}
+
+impl Options<options_typestate::Unauthenticated> {
+    /// Create a new NATS connection. This will not be a connected connection.
     ///
     /// # Example
     /// ```
     /// # fn main() -> std::io::Result<()> {
-    /// let nc = nats::Options::new()
-    ///     .with_name("My App")
-    ///     .connect("demo.nats.io")?;
+    /// let nc = nats::Options::new().connect("demo.nats.io")?;
     /// # Ok(())
     /// # }
     /// ```
-    pub fn with_name(mut self, name: &str) -> Options {
-        self.name = Some(name.to_string());
-        self
+    pub fn new() -> Options<options_typestate::Unauthenticated> {
+        Options::default()
     }
 
-    /// Select option to not deliver messages that we have published.
+    /// Authenticate this NATS connection with a token.
     ///
     /// # Example
     /// ```
     /// # fn main() -> std::io::Result<()> {
     /// let nc = nats::Options::new()
```

Comment:
nats::Options::new() seems like alot to me, you?

Reply:
It's the doc for the configuration type. Doing what the std does is usually the best way to reduce friction for users: https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.new

Reply:
That is fair.